gnat2goto/driver/tree_walk.adb

```diff
@@ -2806,6 +2809,204 @@ package body Tree_Walk is
+         Source_Location => Source_Location);
+      Expon_Greater_Zero : constant Irep := Make_Op_Gt
+        (Lhs => Exponent_Sym,
+         Rhs => Integer_Constant_To_Expr
+           (Value => Uint_0,
+            Expr_Type => Exponent_Type,
+            Source_Location => Source_Location),
+         Source_Location => Source_Location,
+         I_Type => Make_Bool_Type);
+      Set_Expon_Result_To_One : constant Irep := Make_Code_Assign
+        (Lhs => Expon_Result,
+         Rhs => Integer_Constant_To_Expr
+           (Value => Uint_1,
+            Expr_Type => Mod_Type,
+            Source_Location => Source_Location),
+         Source_Location => Source_Location);
+      Square_Base : constant Irep := Make_Code_Assign
+        (Lhs => Base_Sym,
+         Rhs => Do_Operator_Mod
+           (LHS => Base_Sym,
+            Op_Kind => I_Op_Mul,
+            RHS => Base_Sym,
+            Ret_Type => Mod_Type),
+         Source_Location => Source_Location);
+      Halve_Exponent : constant Irep := Make_Code_Assign
+        (Lhs => Exponent_Sym,
+         Rhs => Make_Op_Div
+           (Lhs => Exponent_Sym,
+            Rhs => Integer_Constant_To_Expr
+              (Value => Uint_2,
+               Expr_Type => Exponent_Type,
+               Source_Location => Source_Location),
+            I_Type => Exponent_Type,
+            Source_Location => Source_Location,
+            Div_By_Zero_Check => False),
+         Source_Location => Source_Location);
+      Multiply_Result_With_Base : constant Irep := Make_Code_Assign
+        (Lhs => Expon_Result,
+         Rhs => Do_Operator_Mod
+           (LHS => Expon_Result,
+            Op_Kind => I_Op_Mul,
+            RHS => Base_Sym,
+            Ret_Type => Mod_Type),
+         Source_Location => Source_Location);
+      Decrement_Exponent : constant Irep := Make_Code_Assign
+        (Lhs => Exponent_Sym,
+         Rhs => Make_Op_Sub
+           (Lhs => Exponent_Sym,
+            Rhs => Integer_Constant_To_Expr
+              (Value => Uint_1,
+               Expr_Type => Exponent_Type,
+               Source_Location => Source_Location),
+            I_Type => Exponent_Type,
+            Source_Location => Source_Location),
+         Source_Location => Source_Location);
+      Multiply_Loop_Body : constant Irep := Make_Code_Block
+        (Source_Location => Source_Location);
+      If_Even_Body : constant Irep := Make_Code_Block
+        (Source_Location => Source_Location);
+      If_Odd_Body : constant Irep := Make_Code_Block
+        (Source_Location => Source_Location);
+      Multiply_Loop : constant Irep := Make_Code_While
+        (Loop_Body => Multiply_Loop_Body,
+         Cond => Expon_Greater_Zero,
+         Source_Location => Source_Location);
+      If_Even_Or_Odd_Exponent : constant Irep := Make_Code_Ifthenelse
+        (Cond => Expon_Divisible_By_Two,
+         Then_Case => If_Even_Body,
+         Else_Case => If_Odd_Body,
+         Source_Location => Source_Location);
```

Comment:
Apologies for the long function, but I genuinely can't think of a way to make this shorter. :frowning_face: